llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -21237,6 +21279,32 @@ static SDValue performIntrinsicCombine(SDNode *N,
   switch (IID) {
   default:
     break;
+  case Intrinsic::experimental_vector_partial_reduce_add: {
+    SDLoc DL(N);
+
+    auto NarrowOp = N->getOperand(1);
```

Comment:
This can't just assume that the node is still how we expect (although that will likely almost always be the case). Can you add checks that the nodes are Mul/zext/sext and add some fallback in case they are not?

Reply:
Thanks Dave, I've now added a fallback that creates a chain of ADDs if the nodes aren't as expected.

Reply:
Just realised I got a bit of the logic with `IsValidDotProduct` wrong. I'll fix that up in the next commit.